### ISSUE TITLE
Gradient of  arccos ,arcsin and arctan returns nans when optimization converges.

### ISSUE THREAD

**Issue by miguelamendez** [google/jax#15798]
### Description

What the description says , basically if the computed loss is zero then the gradient of the parameters with respect of the loss returns a vector of nans instead of zeros (as other loss functions do).
This is the code for reproducing the error

import jax
import jax.numpy as jnp
from jax import random

def foo_loss(p,q):
    """Uses arccos also was tested with jax.lax.acos"""
    return jnp.arccos(jnp.vdot(jnp.sqrt(p),jnp.sqrt(q)))

model=jax.nn.softmax

def model_cost(q_logit,p_logit):
    q=model(q_logit)
    p=model(p_logit)
    return foo_loss(p,q)

def model_score(q_logit,p_logit):
    return jax.grad(model_cost)(q_logit,p_logit)

#Example
p_logit=random.normal(random.PRNGKey(0),(4,))
q_logit=random.normal(random.PRNGKey(1),(4,))

print("Diferent vals:",model_cost(q_logit,p_logit),model_score(q_logit,p_logit))
print("Same  vals (after optimization):",model_cost(p_logit,p_logit),model_score(p_logit,p_logit))

### What jax/jaxlib version are you using?

jax 0.4.8

### Which accelerator(s) are you using?

GPU

### Additional system info

python 3.10.6 , ubuntu 22

### NVIDIA GPU info

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:29:00.0  On |                  N/A |
| 33%   37C    P8     5W / 100W |    243MiB /  4096MiB |      3%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1487      G   /usr/lib/xorg/Xorg                147MiB |
|    0   N/A  N/A      3119      G   /usr/bin/gnome-shell               37MiB |
|    0   N/A  N/A      5010      G   ...463712434735781075,131072       55MiB |
+-----------------------------------------------------------------------------+

**Comment by miguelamendez** [comment]
just realised that the derivative of the arcos(1) tends to infinity , a soft fix is to add an eps parameter that adds a small error.

def foo_loss(p,q,eps=1e-6):
"""We don't let it to reach 1"""
return jnp.arccos(jnp.vdot(jnp.sqrt(p),jnp.sqrt(q))-eps)
This partially solves the issue.
